numpy/core/include/numpy/npy_common.h

```diff
@@ -361,49 +363,202 @@ typedef double npy_double;
         compiler and platform
 #endif
-typedef struct { double real, imag; } npy_cdouble;
-
 #if NPY_SIZEOF_COMPLEX_FLOAT != 2 * NPY_SIZEOF_FLOAT
 #error npy_cfloat definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { float real, imag; } npy_cfloat;
-
 #if NPY_SIZEOF_COMPLEX_LONGDOUBLE != 2 * NPY_SIZEOF_LONGDOUBLE
 #error npy_clongdouble definition is not compatible with C99 complex definition ! \
         Please contact NumPy maintainers and give detailed information about your \
         compiler and platform
 #endif
-typedef struct { npy_longdouble real, imag; } npy_clongdouble;
+
+
+#if !defined(__cplusplus) && defined(_MSC_VER) && !defined(__INTEL_COMPILER)
+typedef _Dcomplex npy_cdouble;
+typedef _Fcomplex npy_cfloat;
+typedef _Lcomplex npy_clongdouble;
+#define NPY_CDOUBLE_INIT(real, imag) _CBuild(real, imag)
+#define NPY_CFLOAT_INIT(real, imag) _FCBuild(real, imag)
+#define NPY_CLONGDOUBLE_INIT(real, imag) _LCBuild(real, imag)
+#elif defined(__cplusplus) /* && (!defined(_MSC_VER) || defined(__INTEL_COMPILER)) */
+typedef std::complex<double> npy_cdouble;
+typedef std::complex<float> npy_cfloat;
+typedef std::complex<longdouble_t> npy_clongdouble;
+#define NPY_CDOUBLE_INIT(real, imag) std::complex<double>(real, imag)
+#define NPY_CFLOAT_INIT(real, imag) std::complex<float>(real, imag)
+#define NPY_CLONGDOUBLE_INIT(real, imag) std::complex<longdouble_t>(real, imag)
+#else /* !defined(__cplusplus) && (!defined(_MSC_VER) || defined(__INTEL_COMPILER)) */
+typedef complex double npy_cdouble;
+typedef complex float npy_cfloat;
+typedef complex longdouble_t npy_clongdouble;
+#define NPY_CDOUBLE_INIT(real, imag) CMPLX(real, imag)
+#define NPY_CFLOAT_INIT(real, imag) CMPLXF(real, imag)
+#define NPY_CLONGDOUBLE_INIT(real, imag) CMPLXL(real, imag)
+#endif
+
+#ifndef __cplusplus
+typedef union {
+    double arr[2];
+    npy_cdouble comp;
+} _npy_cdouble_to_arr;
+
+typedef union {
+    float arr[2];
+    npy_cfloat comp;
+} _npy_cfloat_to_arr;
+
+typedef union {
+    longdouble_t arr[2];
+    npy_clongdouble comp;
+} _npy_clongdouble_to_arr;
+#endif
+
+static inline double NPY_CDOUBLE_GET_REAL(const npy_cdouble *c) {
+#ifdef __cplusplus
+    return reinterpret_cast<const double *>(c)[0];
+#else
+    _npy_cdouble_to_arr tmp;
+    tmp.comp = *c;
+    return tmp.arr[0];
+#endif
+}
+
+static inline double NPY_CDOUBLE_GET_IMAG(const npy_cdouble *c) {
+#ifdef __cplusplus
+    return reinterpret_cast<const double *>(c)[1];
+#else
+    _npy_cdouble_to_arr tmp;
+    tmp.comp = *c;
+    return tmp.arr[1];
+#endif
+}
+
+static inline void NPY_CDOUBLE_SET_REAL(npy_cdouble *c, double real) {
+#ifdef __cplusplus
+    double *tmp = reinterpret_cast<double *>(c);
+    tmp[0] = real;
+#else
+    _npy_cdouble_to_arr tmp;
+    tmp.comp = *c;
+    tmp.arr[0] = real;
+    *c = tmp.comp;
+#endif
```

Comment:
I assume the `tmp.arr[0]` turn into a single offset in assembler, but maybe worth a check in [the compiler explorer](https://godbolt.org/)